File: cpp/op.cpp
#include "op.hpp"

// other implementation
/*inline arr Operator::internal_sum(const arr& phi) {*/
    //arr phiC = arr::Zero(phi.rows(), mesh.nInternalCells);
    //for (int j = 0; j < mesh.nFaces; j++) {
        //for (int i = 0; i < phi.rows(); i++) {
            //int p = mesh.owner(0, j);
            //phiC(i, p) += phi(i, j)*mesh.areas(0, j)/mesh.volumes(0, p);
            //if (j < mesh.nInternalFaces) {
                //int n = mesh.neighbour(0, j);
                //phiC(i, n) -= phi(i, j)*mesh.areas(0, j)/mesh.volumes(0, n);
            //}
        //}
    //}
   //return phiC;
//}
//

template<integer shape1>
void Operator::grad(const scalar* phi, arrType<scalar, shape1, 3>& gradPhi, integer index, bool neighbour) {
    const Mesh& mesh = *this->mesh;
    
    integer p = mesh.owner(index);
    integer n = mesh.neighbour(index);
    scalar wp = mesh.areas(index)/mesh.volumes(p);
    scalar wn;
    if (neighbour) {
        wn = mesh.areas(index)/mesh.volumes(n);
    }
    //cout << index << " " << mesh.nInternalFaces << " " << mesh.nFaces << " " << mesh.nInternalCells << endl;
    //cout << p << " " << mesh.areas(index) << " " << mesh.areas.data[index] << endl;
    for (integer i = 0; i < shape1; i++) {
        for (integer j = 0; j < 3; j++) {
            gradPhi(p, i, j) += phi[i]*mesh.normals(index, j)*wp;
            if (neighbour) {
                gradPhi(n, i, j) -= phi[i]*mesh.normals(index, j)*wn;
            }
        }
    }
}
template void Operator::grad(const scalar* phi, arrType<scalar, 1, 3>& gradPhi, integer index, bool neighbour);
template void Operator::grad(const scalar* phi, arrType<scalar, 3, 3>& gradPhi, integer index, bool neighbour);

template<integer shape1>
void Operator::div(const scalar* phi, arrType<scalar, shape1>& divPhi, integer index, bool neighbour) {
    const Mesh& mesh = *this->mesh;
    
    integer p = mesh.owner(index);
    integer n = mesh.neighbour(index);
    scalar wp = mesh.areas(index)/mesh.volumes(p);
    scalar wn;
    if (neighbour) {
        wn = mesh.areas(index)/mesh.volumes(n);
    }
    for (integer i = 0; i < shape1; i++) {
        divPhi(p, i) += phi[i]*wp;
        if (neighbour) {
            divPhi(n, i) -= phi[i]*wn;
        }
    }
}
template void Operator::div(const scalar* phi, arrType<scalar, 1>& divPhi, integer index, bool neighbour);
template void Operator::div(const scalar* phi, arrType<scalar, 3>& divPhi, integer index, bool neighbour);

template<integer shape1>
void Operator::absDiv(const scalar* phi, arrType<scalar, shape1>& divPhi, integer index, bool neighbour) {
    const Mesh& mesh = *this->mesh;
    
    integer p = mesh.owner(index);
    integer n = mesh.neighbour(index);
    scalar wp = mesh.areas(index)/mesh.volumes(p);
    scalar wn;
    if (neighbour) {
        wn = mesh.areas(index)/mesh.volumes(n);
    }
    for (integer i = 0; i < shape1; i++) {
        divPhi(p, i) += phi[i]*wp;
        if (neighbour) {
            divPhi(n, i) += phi[i]*wn;
        }
    }
}
template void Operator::absDiv(const scalar* phi, arrType<scalar, 1>& divPhi, integer index, bool neighbour);

template<integer shape1>
void Operator::snGrad(const arrType<scalar, shape1>& phi, scalar* snGradPhi, integer index) {
    const Mesh& mesh = *this->mesh;
    
    integer p = mesh.owner(index);
    integer n = mesh.neighbour(index);
    integer d = mesh.deltas(index);
    for (integer i = 0; i < shape1; i++) {
        snGradPhi[i] = (phi(n, i)-phi(p, i))*d;
    }
}
template void Operator::snGrad(const arrType<scalar, 1>& phi, scalar* snGradPhi, integer index);
template void Operator::snGrad(const arrType<scalar, 3>& phi, scalar* snGradPhi, integer index);

//arr Operator::div(const arr& phi) {
    //return internal_sum(phi);
//}

//arr Operator::snGrad(const arr& phi) {
    //arr phiN = slice(phi, mesh.neighbour);
    //arr phiP = slice(phi, mesh.owner);
    //return ROWMUL(phiN-phiP, mesh.deltas);
//}

//arr Operator::laplacian(const arr& phi, const arr& DT) {
    //return internal_sum(DT * snGrad(phi));
//}

File: adFVM/op.py
from __future__ import print_function
import numpy as np

from . import config
from .field import Field, CellField, IOField

from adpy.tensor import Tensor

logger = config.Logger(__name__)


def div(phi, mesh, neighbour):
    # needs to be already dotted with mesh normals
    wp = mesh.areas/mesh.volumesL
    if neighbour:
        wn = -mesh.areas/mesh.volumesR
        dphi = Tensor.collate(phi*wp, mesh.owner, phi*wn, mesh.neighbour)
    else:
        dphi = Tensor.collate(phi*wp, mesh.owner)
    return dphi

def absDiv(phi, mesh, neighbour):
    wp = mesh.areas/mesh.volumesL
    if neighbour:
        wn = mesh.areas/mesh.volumesR
        dphi = Tensor.collate(phi*wp, mesh.owner, phi*wn, mesh.neighbour)
    else:
        dphi = Tensor.collate(phi*wp, mesh.owner)
    return dphi


def grad(phi, mesh, neighbour):
    wp = mesh.areas/mesh.volumesL
    if phi.shape == (1,):
        phiN = phi*mesh.normals
    else:
        phiN = phi.outer(mesh.normals)
    if neighbour:
        wn = -mesh.areas/mesh.volumesR
        gphi = Tensor.collate(phiN*wp, mesh.owner, phiN*wn, mesh.neighbour)
    else:
        gphi = Tensor.collate(phiN*wp, mesh.owner)
    return gphi

def gradCell(phi, mesh):
    gradPhi = 0
    nCellFaces = 6
    for i in range(0, nCellFaces):
        P = mesh.cellFaces[i]
        S = mesh.areas.extract(P)
        N = mesh.normals.extract(P)
        w = mesh.weights.extract(P) 
        O = mesh.cellOwner[i]
        N = 2*N*O-N
        w = w + O - 2*w*O
        phiP = phi.extract(mesh.cellNeighbours[i])
        phiF = phi.index()*(-w + 1) + phiP*w
        if phi.shape == (1,):
            gradPhi += phiF*S*N
        else:
            gradPhi += phiF.outer(S*N)
    gradPhi = gradPhi/mesh.volumes
    return gradPhi


def snGrad(phiL, phiR, mesh):
    return (phiR - phiL)/mesh.deltas

def snGradCorr(phiL, phiR, gradPhiF, mesh):
    implicit = (phiR - phiL)/mesh.deltas
    cost = mesh.deltasUnit.dot(mesh.normals)
    if phiL.shape == (1,):
        explicit = gradPhiF[0].dot(mesh.normals-mesh.deltasUnit/cost)
    else:
        explicit = gradPhiF.tensordot(mesh.normals-mesh.deltasUnit/cost)
    return implicit/cost + explicit
   

# code gen ends heere
def internal_sum(phi, mesh, sumOp=False, absolute=False):
    if sumOp:
        if not absolute:
            sumOp = mesh.sumOp
        else:
            sumOp = np.abs(mesh.sumOp)
        #x = (adsparse.basic.dot(sumOp, (phi.field * mesh.areas)))/mesh.volumes
        x = ad.sparse_tensor_dense_matmul(sumOp, phi.field * mesh.areas)/mesh.volumes
    else:
        phiF = phi.field*mesh.areas
        dimensions = (np.product(phi.dimensions),)
        x = np.zeros((mesh.nInternalCells+1,) + dimensions, config.precision)
        np.add.at(x, mesh.owner, phiF)
        if not absolute:
            np.add.at(x, mesh.neighbour[:mesh.nInternalFaces], -phiF[:mesh.nInternalFaces])
        else:
            np.add.at(x, mesh.neighbour[:mesh.nInternalFaces], phiF[:mesh.nInternalFaces])
        x = x[:-1]/mesh.volumes

    return x


def divOld(phi, U=None, ghost=False):
    logger.info('divergence of {0}'.format(phi.name))
    mesh = phi.mesh
    if U is None:
        divField = internal_sum(phi, mesh)
    else:
        assert phi.dimensions == (1,)
        divField = internal_sum((phi*U).dotN(), mesh)
    if ghost:
        divPhi = CellField('div({0})'.format(phi.name), divField, phi.dimensions, ghost=True)
        return divPhi
    else:
        return Field('div({0})'.format(phi.name), divField, phi.dimensions)

def gradOld(phi, op=False, sumOp=False, ghost=False):
    assert len(phi.dimensions) == 1
    logger.info('gradient of {0}'.format(phi.name))
    mesh = phi.mesh
    if phi.dimensions == (1,):
        dimensions = (3,)
    else:
        dimensions = phi.dimensions + (3,)

    if op:
        gradField = adsparse.basic.dot(mesh.gradOp, phi.field)
        gradField = gradField.reshape((mesh.nInternalCells,) + dimensions)
        if dimensions == (3,3):
            gradField = gradField.transpose((0, 2, 1))
    else:
        N = Field('N', mesh.normals, (3,))
        if dimensions == (3,):
            product = phi * N
        else:
            product = phi.outer(N)
            dimprod = np.prod(dimensions)
            product.field = np.reshape(product.field, (mesh.nFaces, dimprod))
        gradField = internal_sum(product, mesh, sumOp=sumOp)
        # if grad of vector
        if len(dimensions) == 2:
            gradField = np.reshape(gradField, (mesh.nInternalCells,) + dimensions)

    return IOField('grad({0})'.format(phi.name), gradField, dimensions, ghost=ghost)

